File: android/gemcore/src/main/kotlin/com/wallet/core/primitives/Staking.kt
package com.wallet.core.primitives

data class RedelegateData(
    val delegation: Delegation,
    val toValidator: DelegationValidator,
)

sealed class StakeType {
    data class Stake(val content: DelegationValidator) : StakeType()
    data class Unstake(val content: Delegation) : StakeType()
    data class Redelegate(val content: RedelegateData) : StakeType()
    data class Rewards(val content: List<DelegationValidator>) : StakeType()
    data class Withdraw(val content: Delegation) : StakeType()
    data class Freeze(val content: Resource) : StakeType()
    data class Unfreeze(val content: Resource) : StakeType()
}

File: ios/Packages/Primitives/Sources/Staking.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation

public struct RedelegateData: Codable, Equatable, Hashable, Sendable {
    public let delegation: Delegation
    public let toValidator: DelegationValidator

    public init(delegation: Delegation, toValidator: DelegationValidator) {
        self.delegation = delegation
        self.toValidator = toValidator
    }
}

public enum StakeType: Codable, Equatable, Hashable, Sendable {
    case stake(DelegationValidator)
    case unstake(Delegation)
    case redelegate(RedelegateData)
    case rewards([DelegationValidator])
    case withdraw(Delegation)
    case freeze(Resource)
    case unfreeze(Resource)
}
